File: static/ts/trivia.ts
interface Comparable<T> {
    
    compareTo(t: T): number;
    
}

interface Ordered<T> {
    
    readonly order: (a: T, b: T) => number;
    
}

interface IdAble {
    
    readonly id: number;
    
}

interface Named {
    
    readonly name: string;
    
}

class NamedImpl implements Named {
    
    public readonly name: string;
    
    public constructor(name: string) {
        this.name = name;
    }
    
}

const names = function(names: string[]): Named[] {
    return names.map(name => new NamedImpl(name));
};

interface Properties<T> {
    
    [key: string]: T;
    
}

interface EnumCtor<CtorArgs> {
    
    new(args?: CtorArgs): any;
    
}

interface IEnum<T> {
    
    readonly values: T[];
    
    get(name: string): T | undefined;
    
    getNonNull(name: string): T;
    
}

class EmptyEnum {
    
    public constructor(args?: any) {}
    
}

const newEnum = function <T, CtorArgs>(enumProto: EnumCtor<CtorArgs>, argsList: (Named & CtorArgs)[]) {
    return class Enum extends enumProto implements Comparable<T>, Ordered<T> {
        
        private static createValueMap(): Map<string, T & Enum> {
            const map: Map<string, T & Enum> = new Map();
            Enum.values.forEach(value => {
                map.set(value.name, value);
            });
            return map;
        }
        
        public static readonly values: (T & Enum)[] =
            argsList.map((args: Named & CtorArgs, i: number) => <T & Enum> new Enum(i, args));
        
        private static readonly instances: Map<string, T & Enum> = Enum.createValueMap();
        
        public static getNullable(name: string): (T & Enum) | undefined {
            return Enum.instances.get(name);
        }
        
        public static get(name: string): T & Enum {
            return <T & Enum> Enum.getNullable(name);
        }
        
        public readonly id: number;
        public readonly name: string;
        
        public constructor(id: number, args: Named & CtorArgs) {
            super(<CtorArgs> args);
            this.id = id;
            this.name = args.name;
        }
        
        public compareTo(other: T & Enum): number {
            return this.id - other.id;
        }
        
        public order: (a: T & Enum, b: T & Enum) => number =
            (a: T & Enum, b: T & Enum) => a.id - b.id;
        
    }
};

class Category extends newEnum<Category, any>(EmptyEnum, names([
    "multiple",
    "boolean"
])) {
    
    public static readonly MultipleChoice: Category = Category.get("multiple");
    
    public static readonly TrueOrFalse: Category = Category.get("boolean");
    
}

class QuestionType extends newEnum<QuestionType, any>(EmptyEnum, names([
    "easy",
    "medium",
    "hard"
])) {
    
    public static readonly Easy: Category = Category.get("easy");
    
    public static readonly Medium: Category = Category.get("medium");
    
    public static readonly Hard: Category = Category.get("hard");
    
}

class Difficulty extends newEnum<Difficulty, any>(EmptyEnum, names([
    "Any Category",
    "General Knowledge",
    "Entertainment: Books",
    "Entertainment: Film",
    "Entertainment: Music",
    "Entertainment: Musicals & Theatres",
    "Entertainment: Television",
    "Entertainment: Video Games",
    "Entertainment: Board Games",
    "Science & Nature",
    "Science: Computers",
    "Science: Mathematics",
    "Mythology",
    "Sports",
    "Geography",
    "History",
    "Politics",
    "Art",
    "Celebrities",
    "Animals",
    "Vehicles",
    "Entertainment: Comics",
    "Science: Gadgets",
    "Entertainment: Japanese Anime & Manga",
    "Entertainment: Cartoon & Animations"
])) {
    
    public static readonly Any_Category: QuestionType =
            QuestionType.get("Any Category");

    public static readonly General_Knowledge: QuestionType =
            QuestionType.get("General Knowledge");

    public static readonly Entertainment_Books: QuestionType =
            QuestionType.get("Entertainment: Books");

    public static readonly Entertainment_Film: QuestionType =
            QuestionType.get("Entertainment: Film");

    public static readonly Entertainment_Music: QuestionType =
            QuestionType.get("Entertainment: Music");

    public static readonly Entertainment_Musicals_Theatres: QuestionType =
            QuestionType.get("Entertainment: Musicals & Theatres");

    public static readonly Entertainment_Television: QuestionType =
            QuestionType.get("Entertainment: Television");

    public static readonly Entertainment_Video_Games: QuestionType =
            QuestionType.get("Entertainment: Video Games");

    public static readonly Entertainment_Board_Games: QuestionType =
            QuestionType.get("Entertainment: Board Games");

    public static readonly Science_Nature: QuestionType =
            QuestionType.get("Science & Nature");

    public static readonly Science_Computers: QuestionType =
            QuestionType.get("Science: Computers");

    public static readonly Science_Mathematics: QuestionType =
            QuestionType.get("Science: Mathematics");

    public static readonly Mythology: QuestionType =
            QuestionType.get("Mythology");

    public static readonly Sports: QuestionType =
            QuestionType.get("Sports");

    public static readonly Geography: QuestionType =
            QuestionType.get("Geography");

    public static readonly History: QuestionType =
            QuestionType.get("History");

    public static readonly Politics: QuestionType =
            QuestionType.get("Politics");

    public static readonly Art: QuestionType =
            QuestionType.get("Art");

    public static readonly Celebrities: QuestionType =
            QuestionType.get("Celebrities");

    public static readonly Animals: QuestionType =
            QuestionType.get("Animals");

    public static readonly Vehicles: QuestionType =
            QuestionType.get("Vehicles");

    public static readonly Entertainment_Comics: QuestionType =
            QuestionType.get("Entertainment: Comics");

    public static readonly Science_Gadgets: QuestionType =
            QuestionType.get("Science: Gadgets");

    public static readonly Entertainment_Japanese_Anime_Manga: QuestionType =
            QuestionType.get("Entertainment: Japanese Anime & Manga");

    public static readonly Entertainment_Cartoon_Animations: QuestionType =
            QuestionType.get("Entertainment: Cartoon & Animations");
    
}

class Question {
    
    readonly category: Category;
    readonly type: QuestionType;
    readonly difficulty: Difficulty;
    readonly answers: Array<string>;
    readonly correctAnswer: string;
    
    public constructor(json: any) {
        this.category = Category.get(json.category);
        this.type = QuestionType.get(json.type);
        this.difficulty = Difficulty.get(json.difficulty);
        this.correctAnswer = json.correct_answer;
        this.answers = [this.correctAnswer];
        this.answers.push(json.incorrect_answers);
    }
    
}

File: api/trivia.py
#!usr/bin/python
import requests, json, sqlite3
from HTMLParser import HTMLParser

#Corresponding Categories#
#PLEASE DONT TOUCH IT#
cat = [
 0,0,0,0,0,0,0,0,0,
 [9, 'General Knowledge'],
 [10, 'Entertainment: Books'],
 [11, 'Entertainment: Film'],
 [12, 'Entertainment: Music'],
 [13, 'Entertainment: Musicals & Theatres'],
 [14, 'Entertainment: Television'],
 [15, 'Entertainment: Video Games'],
 [16, 'Entertainment: Board Games'],
 [17, 'Science & Nature'],
 [18, 'Science: Computers'],
 [19, 'Science: Mathematics'],
 [20, 'Mythology'],
 [21, 'Sports'],
 [22, 'Geography'],
 [23, 'History'],
 [24, 'Politics'],
 [25, 'Art'],
 [26, 'Celebrities'],
 [27, 'Animals'],
 [28, 'Vehicles'],
 [29, 'Entertainment: Comics'],
 [30, 'Science: Gadgets'],
 [31, 'Entertainment: Japanese Anime & Manga'],
 [32, 'Entertainment: Cartoon & Animations']
]
typearr = ["", "multiple", "boolean"]
diffarr = ["", "easy", "medium", "hard"]


'''
This will return a list of dictionaries of questions in this format

[{"question":___,"difficulty":___,"category":___,correct_ans":___,"wrong_ans":[__,__,__],"type":___ },]

Guide to arguments:

int amount - amount of questions to request, do not exceed 50.

--The args below are optional--
int typing - 1:multiple, 2:boolean
int category - *see bottom of file for corresponding categories
int difficulty - 1:easy, 2:med, 3:hard
'''

def get_write_witcha(amount, typing=0, category=0, difficulty=0):
    data = get_questions(amount, typing=0, category=0, difficulty=0)
    db_name = "../data/listen_up.db"
    dab = sqlite3.connect(db_name)
    c = dab.cursor()
    for item in data:
        cmd = """SELECT question FROM questions WHERE question = "%s" """%(item["question"])
        c.execute(cmd)
        questionable = c.fetchone()
        if questionable == None:        
            cmd = "SELECT MAX(id) FROM questions"
            c.execute(cmd)
            newID = c.fetchone()[0]
            if newID == None:
            	newID = 0
            else:
            	newID+=1
            	'''
            cmd = "INSERT INTO question values(%i,%s,%s,%s,%i,%i,%s,%s)"%(
            	newID,
            	item["question"],
            	item["correct_answer"],
            	''.join(item["incorrect_answers"]),
            	typing, mult or bool?
            	difficulty, easy, med, hard?
			'''
            print "theres nothing here"
        else:
            print "you screwed up"
    return data



def get_questions(amount, typing=0, category=0, difficulty=0):
    querystr = query_help(amount, typing, category, difficulty)
    source = "https://opentdb.com/api.php?" + querystr
    datastuff = requests.get(source)
    datastuff = datastuff.json()
    #Ew HTML Encoding
    h = HTMLParser()
    results = datastuff["results"]
    for item in results:
        keys = item.keys()
        for key in keys:
            item[key] = h.unescape(item[key])
            try:
                item[key] = item[key].replace("\"","'")
            except(Exception):
                '''
                print item["question"]
                print item["correct_answer"]
                print item["incorrect_answers"]
                '''
                pass
    print results
    return results


'''
query_help will help return a query string for things above 
'''


def query_help(amount, typing, category, difficulty):
    retstring = ""
    retstring += "amount=" + str(amount)
    if typing != 0: retstring += "&type=" + typearr[typing]
    if category < 9: retstring += "&category=" + str(category)
    if difficulty != 0: retstring += "&difficulty=" + diffarr[difficulty]
    return retstring

if __name__ == '__main__':
    get_write_witcha(1,0,9,2)
